0000-private-event.md

```diff
@@ -0,0 +1,146 @@
+---
+layout: default
+title: Private Chaincode Events
+nav_order: 3
+---
+
+- Feature Name: private_chaincode_events
+- Start Date: 2025-04-22
+- RFC PR:
+- Fabric Component: core, peer
+- Fabric Issue:
+
+# Summary
+[summary]: #summary
+
+This RFC introduces the concept of **Private Chaincode Events** for Hyperledger Fabric. These events are generated by chaincode but differ from existing chaincode events in that they are stored and disseminated only to explicitly specified organizations and identities, without ever being recorded on the public ledger. This feature enhances privacy and fine-grained control over event visibility.
+
+# Motivation
+[motivation]: #motivation
+
+Several Hyperledger Fabric users have expressed the need to emit chaincode events that are:
+
+1. Invisible to unauthorized parties.
+2. Not stored on the public ledger.
+3. Delivered only to explicitly listed organizations and users.
+4. Replayable from a specific block or timestamp.
+
+Current public chaincode events cannot satisfy these privacy requirements. This proposal aims to provide a solution by building on existing private data collections (PDCs) and enhancing the event infrastructure to support controlled dissemination of private events.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Concept: Private Chaincode Events
+
+Private Chaincode Events are triggered by invoking the standard `PutPrivateData()` method in chaincode, with a specially prefixed key (e.g., `__hlf_privateEvent__eventName`).
+
+Each private event must conform to a strict structure:
+
+```json
+{
+  "visibleTo": {
+    "Org1MSP": ["user1"],
+    "Org2MSP": ["user2", "user3"]
+  },
+  "payload": {
+    "from": "A",
+    "to": "B",
+    "amount": 87
+  }
+}
+```
+
+The peer recognizes this as a private event and tracks it for notification purposes.
+
+## Listening to Private Events
+
+The private events are restricted and stored in the similar way as the private data.
+
+Only peers in organizations and CNs listed in the event's `visibleTo` map will process and relay the event to their REST Proxy, which will forward it to eligible subscribers.
+
+## Replaying Events
+
+Private events are replay-able by leveraging their PDC storage. Even though they are not part of the public ledger, the peer can match them to their corresponding transaction ID and block, making replay feasible from a given block number.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Event Creation
+- A chaincode uses `PutPrivateData(<pdc>, "__hlf_privateEvent__<name>", <eventData>)` to raise a private event.
+- `eventData` must be valid JSON with the `visibleTo` and the `payload` fields.
```

Comment:
If you would like to actually introduce a new capability, doesn't it deserve APIs on its own?

Reply:
It's possible, will update.